compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -2133,13 +2133,14 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                         self.current -= 1;
                     }
                     fn visit_expr(&mut self, expr: &hir::Expr<'tcx>) {
-                        if self.span == expr.span {
+                        if self.span == expr.span.source_callsite() {
```

Comment:
Can you explain what this is actually doing? It's not immediately clear what the failure mode was previously before this PR.

Side-note, I wonder if this should be using `Span::find_ancestor_inside` instead of `source_callsite`, perhaps to deal with multiply nested macro exprs.

Reply:
`source_callsite` is recursive as well, from what I could tell.

When you have `let x = vec![1, 2, 3].iter();`, the span in the obligation corresponds to the expression inside the macro, which points into stdlib. By calling `source_callsite` we get the span to `vec![1, 2, 3]`. We then also account for it in the visitor. The reason it wasn't found before was mainly because we also checked that the span we saw was within the statement, if it wasn't we just bailed.

Reply:
This may be purely academic, but the usage of `Span::*_callsite` functions (which are absolute, in a sense) instead of `Span::find_ancestor_*` (which are relative) means that we somewhat arbitrarily provide poorer suggestions for code like:

```rust
macro_rules! foo {
    () => {
        fn main() {
            let mut x = vec![1].iter();
            x.next();
        }
    };
}

foo!();
```

But whatever I guess. Just wanted to note that using `source_callsite`/etc is a *tiny* bit sketchy.

Reply:
Ah I see. The fact that it recursively climbs is the problem you wanted to highlight. That's fair. I was thinking of the opposite case, where lets say `vec` is actually nested a few macros deep before actually expanding to an expression, where we wouldn't suggest the binding. I guess either one or the other works, until we start properly tracking the expression instead of trying to reconstruct things from a stray span.